crates/primitives-traits/src/transaction/mod.rs

```diff
@@ -1,28 +1,65 @@
-//! Transaction abstraction
+use core::{fmt::Debug, hash::Hash};
 
-pub mod signed;
-
-use alloc::fmt;
+use alloy_primitives::{TxKind, B256};
 
+use alloy_rlp::Encodable;
 use reth_codecs::Compact;
 use serde::{Deserialize, Serialize};
 
-/// Helper trait that unifies all behaviour required by transaction to support full node operations.
-pub trait FullTransaction: Transaction + Compact {}
-
-impl<T> FullTransaction for T where T: Transaction + Compact {}
+pub mod signed;
 
-/// Abstraction of a transaction.
-pub trait Transaction:
-    alloy_consensus::Transaction
+#[allow(dead_code)]
+/// Inner trait for a raw transaction.
+pub trait Transaction<T>:
+    Debug
+    + Default
     + Clone
-    + fmt::Debug
-    + PartialEq
     + Eq
-    + Default
-    + alloy_rlp::Encodable
-    + alloy_rlp::Decodable
+    + PartialEq
+    + Encodable
+    + Hash
     + Serialize
     + for<'de> Deserialize<'de>
+    + From<T>
+    + alloy_consensus::Transaction
 {
+    /// Heavy operation that return signature hash over rlp encoded transaction.
+    /// It is only for signature signing or signer recovery.
+    fn signature_hash(&self) -> B256;
+
+    /// Gets the transaction's [`TxKind`], which is the address of the recipient or
+    /// [`TxKind::Create`] if the transaction is a contract creation.
+    fn kind(&self) -> TxKind;
+
+    /// Returns true if the tx supports dynamic fees
+    fn is_dynamic_fee(&self) -> bool;
+
+    /// Returns the effective gas price for the given base fee.
+    fn effective_gas_price(&self, base_fee: Option<u64>) -> u128;
+
+    /// This encodes the transaction _without_ the signature, and is only suitable for creating a
+    /// hash intended for signing.
+    fn encode_without_signature(&self, out: &mut dyn bytes::BufMut);
+
+    /// Calculates a heuristic for the in-memory size of the [Transaction].
+    fn size(&self) -> usize;
 }
+
+/// Trait that defines methods of a raw transaction.
+///
+/// Transaction types were introduced in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718).
+#[cfg(feature = "arbitrary")]
+pub trait FullTransaction<T>: Transaction<T> + Compact + for<'b> arbitrary::Arbitrary<'b> {}
```

Comment:
when asked to revert, please also revert the code docs. one should always think, that the docs and the code are glued together and are inseparable.
```suggestion
/// Helper trait that unifies all behaviour required by transaction to support full node operations.
#[cfg(feature = "arbitrary")]
pub trait FullTransaction<T>: Transaction<T> + Compact + for<'b> arbitrary::Arbitrary<'b> {}
```